openmc/tallies.py

```diff
@@ -1487,10 +1487,16 @@ def _outer_product(self, other, binary_op):
         other_filters = set(other.filters)
         filter_intersect = self_filters.intersection(other_filters)
 
-        # Align the shared filters to follow in each tally operand
+        # Align the shared filters in successive order
         for i, filter in enumerate(filter_intersect):
-            self_index = self.filters.index(filter)
-            other_filter = other.filters[self_index]
+            self_filter = self.filters[i]
+            other_filter = other.filters[i]
+
+            # If necessary, swap self filter
+            if self_filter != filter:
+                self = self.swap_filters(filter, self_filter)
```

Comment:
It seems odd to set `self` here, is this what you intended? I'm not even sure what this will do in Python.